### PR TITLE
Implement `Selector` (#574 part1)

### PR DESCRIPTION
**Pull request by inikulin**
I wanted to do it one PR, but it's already huge. You can see what's done so far in #574.

\cc @AlexanderMoskovkin @helen-dikareva @VasilyStrelyaev @AndreyBelym 

Guys, please review this ASAP because I'm blocking on this PR.

### CODE REVIEW

File: src/compiler/es-next/hybrid-function.js [hunk 1]
```diff
@@ -122,8 +122,7 @@ function compile (fnCode, dependenciesCode, createRegeneratorInClientCodeError)
     if (NODE_VER >= 4)
         fnCode = downgradeES(fnCode);
 
-    if (!hammerhead.isScriptProcessed(fnCode))
-        fnCode = hammerhead.processScript(fnCode, false);
+    fnCode = hammerhead.processScript(fnCode, false);
```
Comment (AndreyBelym) [inline src/compiler/es-next/hybrid-function.js]: Is this change intentional?

File: src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        [
+            'namespaceURI', 'id',
+            'value', 'checked',
+            'scrollWidth', 'scrollHeight', 'scrollLeft', 'scrollTop',
+            'offsetWidth', 'offsetHeight', 'offsetLeft', 'offsetTop',
+            'clientWidth', 'clientHeight', 'clientLeft', 'clientTop'
+        ].forEach(prop => this[prop] = element[prop]);
+    }
+
+    static _getBoundingClientRect (element) {
```
Comment (helen-dikareva) [inline src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js]: Isn't it expected returns object not like DOM method `getBoundingClientRect`? https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIDOMClientRect

File: src/client/driver/command-executors/execute-hybrid-function/index.js [hunk 3]
```diff
@@ -2,51 +2,71 @@ import hammerhead from '../../deps/hammerhead';
 import DriverStatus from '../../status';
 import Replicator from 'replicator';
 import evalFunction from './eval-function';
-import { UncaughtErrorInClientExecutedCode } from '../../../../errors/test-run';
-
-const HYBRID_COMPILED_CODE = '[[hybridCompiledCode]]';
+import { NodeSnapshot, ElementSnapshot } from './node-snapshots';
+import { UncaughtErrorInHybridFunctionCode, DomNodeHybridResultError } from '../../../../errors/test-run';
 
+// NOTE: save original ctors because they may be overwritten by use code
```
Comment (VasilyStrelyaev) [inline src/client/driver/command-executors/execute-hybrid-function/index.js]: user code?
```
// NOTE: save original ctors because they may be overwritten by user code
```

File: src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        this.hasChildElements  = !!this.childElementCount;
+    }
+
+    static _getChildElementCount (node) {
+        var children = getChildren(node);
+
+        if (children)
+            return children.length;
+
+        // NOTE: doesn't have `children` for non-element nodes =/
```
Comment (VasilyStrelyaev) [inline src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js]:
```
        // NOTE: for non-element nodes, which don't have 'children' =/
```

File: test/functional/fixtures/api/es-next/eval/test.js [hunk 5]
```diff
@@ -24,15 +24,15 @@ describe('[API] t.eval', function () {
     it('Should have the correct callsite if an error occurs on instantiation', function () {
         return runTests('./testcafe-fixtures/eval-test.js', 'Error on instantiation', { shouldFail: true })
             .catch(function (errs) {
-                expect(errs[0]).contains('Client code is expected to be specified as a function, but "string" was passed.');
+                expect(errs[0]).contains('Hybrid function code is expected to be specified as a function, but "string" was passed.');
```
Comment (helen-dikareva) [inline test/functional/fixtures/api/es-next/eval/test.js]: I like more previous text for error `'Client code is expected to be specified as a function, but "string" was passed.'`
Because now it look strange, we no longer have term 'Hybrid ' in API. I think we should not use it.
May be we can get rid of `Hybrid ` using message `Code executed on client` for Selector and Client function
Or some other way

File: src/compiler/es-next/hybrid-function.js (hunk 1 above)
Comment (inikulin) [inline src/compiler/es-next/hybrid-function.js]: Yes, we don't have functions as return value for hybrids. So there is no way compiled code can get here anymore

File: src/client/driver/command-executors/execute-hybrid-function/index.js (hunk 3 above)
Comment (inikulin) [inline src/client/driver/command-executors/execute-hybrid-function/index.js]: Page code

File: src/errors/runtime/message.js [hunk 6]
```diff
@@ -22,9 +22,9 @@ export default {
     testBodyIsNotAFunction:             'The test body is expected to be a function, but it was "{type}".',
     beforeEachIsNotAFunction:           `Fixture's "beforeEach" method takes a function, but "{type}" was passed.`,
     afterEachIsNotAFunction:            `Fixture's "afterEach" method takes a function, but "{type}" was passed.`,
-    clientCodeIsNotAFunction:           'Client code is expected to be specified as a function, but "{type}" was passed.',
+    hybridFunctionCodeIsNotAFunction:   'Hybrid function code is expected to be specified as a function, but "{type}" was passed.',
```
Comment (VasilyStrelyaev) [inline src/errors/runtime/message.js]: Can we pass the function type here, so that it becomes

```
    hybridFunctionCodeIsNotAFunction:   '{functionType} function code is expected to be specified as a function, but "{type}" was passed.'
```

File: src/errors/runtime/message.js [hunk 7]
```diff
@@ -22,9 +22,9 @@ export default {
...
     hybridFunctionCantResolveTestRun:   "The hybrid function cannot implicitly resolve the test run in context of which it should be executed. If you need to call the hybrid function from the Node.js API callback, pass the test controller manually via hybrid function's `.bindTestRun(t)` method first. Note that you cannot execute hybrid functions outside the test code.",
-    regeneratorInClientCode:            'Code executed on the client cannot contain generators or `async/await` syntax (use Promises instead).',
+    regeneratorInHybridFunctionCode:    'Hybrid function code cannot contain generators or `async/await` syntax (use Promises instead).',
```
Comment (VasilyStrelyaev) [inline src/errors/runtime/message.js]: And here?

File: src/errors/test-run/templates.js [hunk 8]
```diff
@@ -53,16 +53,16 @@ export default {
         ${err.getCallsiteMarkup()}
     `),
 
-    [TYPE.uncaughtErrorInClientExecutedCode]: err => markup(err, `
-        An error occurred in code executed on the client:
+    [TYPE.uncaughtErrorInHybridFunctionCode]: err => markup(err, `
+        An error occurred in hybrid function code:
```
Comment (VasilyStrelyaev) [inline src/errors/test-run/templates.js]: and here

File: src/errors/test-run/templates.js [hunk 9]
```diff
@@ -53,16 +53,16 @@ export default {
...
 
         <code>${escapeHtml(err.errMsg)}</code>
 
         ${err.getCallsiteMarkup()}
     `),
 
-    [TYPE.clientCodeExecutionInterruptionError]: err => markup(err, `
-        Client code execution was interrupted by page unload. This problem may appear if you trigger page navigation from client code.
+    [TYPE.hybridFunctionExecutionInterruptionError]: err => markup(err, `
+        Hybrid function execution was interrupted by page unload. This problem may appear if you trigger page navigation from hybrid function code.
```
Comment (VasilyStrelyaev) [inline src/errors/test-run/templates.js]: and here

File: src/errors/test-run/templates.js [hunk 10]
```diff
@@ -218,8 +215,14 @@ export default {
     `),
 
     [TYPE.regeneratorInFunctionArgumentOfHybridFunctionError]: err => markup(err, `
-        Hybrid function argument is a function that contains either generators or the <code>async/await</code> syntax. These features cannot be used in client code. Use Promises instead.
+        Hybrid function argument is a function that contains either generators or the <code>async/await</code> syntax. These features cannot be used in hybrid function code. Use Promises instead.
```
Comment (VasilyStrelyaev) [inline src/errors/test-run/templates.js]: and here

File: src/errors/test-run/templates.js [hunk 11]
```diff
@@ -218,8 +215,14 @@ export default {
...
 
         ${err.getCallsiteMarkup()}
+    `),
+
+    [TYPE.domNodeHybridResultError]: err => markup(err, `
+       Regular Hybrid functions cannot return DOM elements. Use <code>Selector</code> functions for this purpose.
```
Comment (VasilyStrelyaev) [inline src/errors/test-run/templates.js]: Client functions cannot return DOM elements. Use <code>Selector</code> functions for this purpose.

File: test/functional/fixtures/api/es-next/eval/test.js (hunk 5 above)
Comment (inikulin) [inline test/functional/fixtures/api/es-next/eval/test.js]: I will take care of errors later (see `Error text adjustment for Hybrids and Selector` in #574 ). Otherwise this PR will have enormous size.

File: src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js (hunk 2 above)
Comment (inikulin) [inline src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js]: Right!

File: src/errors/runtime/message.js (hunk 6 above)
Comment (inikulin) [inline src/errors/runtime/message.js]: https://github.com/DevExpress/testcafe/pull/583#discussion_r66263947

File: test/functional/fixtures/api/es-next/eval/test.js (hunk 5 above)
Comment (helen-dikareva) [inline test/functional/fixtures/api/es-next/eval/test.js]: Ok

File: src/hybrid-function/selector.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+export default class SelectorHybridFunction extends HybridFunction {
+    constructor (fn, dependencies, boundTestRun, callsiteNames) {
+        super(fn, dependencies, boundTestRun, callsiteNames);
+    }
+
+    _getFnCode (fn) {
+        var fnType = typeof fn;
+
+        // TODO needs it's own error and should accepts strings
```
Comment (VasilyStrelyaev) [inline src/hybrid-function/selector.js]: Ignore this if that TODO is here until the next PR.

Otherwise,

```
        // TODO needs its own error and should accepts strings
```

File: src/hybrid-function/selector.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        if (fnType !== 'function')
+            throw new APIError(this.callsiteNames.instantiation, MESSAGE.hybridFunctionCodeIsNotAFunction, fnType);
+
+        return fn.toString();
+    }
+
+    _createExecutionTestRunCommand (args) {
+        // TODO needs it's own command
```
Comment (VasilyStrelyaev) [inline src/hybrid-function/selector.js]: Ignore this if that TODO is here until the next PR.

Otherwise,
```
        // TODO needs its own command
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+    it('Should accept arguments', function () {
+        return runTests('./testcafe-fixtures/client-fn-test.js', 'Call with arguments');
+    });
+
+    it('Should perform Hammerhead code instrumentation on function code', function () {
+        return runTests('./testcafe-fixtures/client-fn-test.js', 'Hammerhead code instrumentation');
+    });
+
+    it('Should raise error if ClientFunction argument is not a function', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should raise an error if the ClientFunction argument is not a function', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+        }).catch(function (errs) {
+            expect(errs[0].indexOf(
+                'Hybrid function code is expected to be specified as a function, but "number" was passed.'
+            )).eql(0);
+
+            expect(errs[0]).contains('> 29 |    await ClientFunction(123)();');
+        });
+    });
+
+    it('Should raise error if ClientFunction function not able to resolve test run', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should raise an error if the ClientFunction's function is unable to resolve test run', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+        }).catch(function (errs) {
+            expect(errs[0].indexOf(
+                'The hybrid function cannot implicitly resolve the test run in context of which it should be executed.'
+            )).eql(0);
+
+            expect(errs[0]).contains(' > 40 |                fn();');
+        });
+    });
+
+    it('Should raise error if ClientFunction function contains async/await syntax', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should raise an error if the ClientFunction's function contains async/await syntax', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+        }).catch(function (errs) {
+            expect(errs[0].indexOf(
+                'Hybrid function code cannot contain generators or `async/await` syntax (use Promises instead).'
+            )).eql(0);
+
+            expect(errs[0]).contains('> 51 |    ClientFunction(async () => Promise.resolve());');
+        });
+    });
+
+    it('Should raise error if ClientFunction function contains generator', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should raise an error if the ClientFunction's function contains a generator', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+        }).catch(function (errs) {
+            expect(errs[0].indexOf(
+                'Hybrid function code cannot contain generators or `async/await` syntax (use Promises instead).'
+            )).eql(0);
+
+            expect(errs[0]).contains('> 55 |    ClientFunction(function*() { ');
+        });
+    });
+
+    it('Should be able to bind test run using `.bindTestRun(t)` method', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should be able to bind a test run using the `.bindTestRun(t)` method', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+        return runTests('./testcafe-fixtures/client-fn-test.js', 'Bind ClientFunction function', { only: 'chrome' });
+    });
+
+    it('Should raise error if ClientFunction bound to non-TestController object', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should raise an error if a ClientFunction is bound to a non-TestController object', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+
+    it('Should support Promises as a result', function () {
+        return runTests('./testcafe-fixtures/client-fn-test.js', 'Promises support');
+    });
+
+    it('Should polyfill Babel artifacts', function () {
+        return runTests('./testcafe-fixtures/client-fn-test.js', 'Babel artifacts polyfills');
+    });
+
+    it('Should handle error in ClientFunction code', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should handle errors in ClientFunction code', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+        return runTests('./testcafe-fixtures/client-fn-test.js', 'Error in code', { shouldFail: true })
+            .catch(function (errs) {
+                expect(errs[0]).contains('An error occurred in hybrid function code:');
+                expect(errs[0]).contains('Error: Hey ya!');
+                expect(errs[0]).contains('> 123 |    await fn();');
+            });
+    });
+
+    it('Should handle error in Promise in ClientFunction code', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should handle errors in a Promise in ClientFunction code', function () {
```

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+                );
+                expect(errs[0]).contains(' > 202 |    await hfn(async () => Promise.resolve());');
+            });
+    });
+
+    it('Should accept a hybrid as an argument', function () {
+        return runTests('./testcafe-fixtures/client-fn-test.js', 'ClientFunction with hybrid argument');
+    });
+
+    it('Should raise error if DOM node is returned', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should raise an error if a DOM node is returned', function () {
```

File: test/functional/fixtures/api/es-next/selector/test.js [hunk 23]
```diff
@@ -0,0 +1,23 @@
+describe('[API] Selector', function () {
+    it('Should provide basic properties in HTMLElement snapshot', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/selector/test.js]:
```
    it('Should provide basic properties in HTMLElement snapshots', function () {
```

File: test/functional/fixtures/api/es-next/selector/test.js [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+        return runTests('./testcafe-fixtures/selector-test.js', 'HTMLElement snapshot basic properties');
+    });
+
+    it('Should provide basic properties in SVGElement snapshot', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/selector/test.js]:
```
    it('Should provide basic properties in SVGElement snapshots', function () {
```

File: test/functional/fixtures/api/es-next/selector/test.js [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+        return runTests('./testcafe-fixtures/selector-test.js', 'SVGElement snapshot basic properties');
+    });
+
+    it('Should provide input-specific properties in element snapshot', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/selector/test.js]:
```
    it('Should provide input-specific properties in element snapshots', function () {
```

File: test/functional/fixtures/api/es-next/selector/test.js [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+        return runTests('./testcafe-fixtures/selector-test.js', 'Input-specific element snapshot properties');
+    });
+
+    it('Should provide `innerText` property in element snapshot', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/selector/test.js]:
```
    it('Should provide the `innerText` property in element snapshots', function () {
```

File: test/server/compiler-test.js [hunk 27]
```diff
@@ -525,8 +526,8 @@ describe('Compiler', function () {
                 });
         });
 
-        it('Should raise error if Hybrid argument is not a function', function () {
-            var testfile = resolve('test/server/data/test-suites/hybrid-arg-is-not-a-function/testfile.js');
+        it('Should raise error if ClientFunction argument is not a function', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        it('Should raise an error if the ClientFunction argument is not a function', function () {
```

File: test/server/compiler-test.js [hunk 28]
```diff
@@ -552,8 +553,8 @@ describe('Compiler', function () {
                 });
         });
 
-        it('Should raise error if Hybrid argument is not a function (if called as ctor)', function () {
-            var testfile = resolve('test/server/data/test-suites/hybrid-arg-is-not-a-function-as-ctor/testfile.js');
+        it('Should raise error if ClientFunction argument is not a function (if called as ctor)', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        it('Should raise an error if the ClientFunction argument is not a function (if called as ctor)', function () {
```

File: test/server/compiler-test.js [hunk 29]
```diff
@@ -580,8 +581,8 @@ describe('Compiler', function () {
         });
 
 
-        it('Should raise error if Hybrid function not able to resolve test run', function () {
-            var testfile = resolve('test/server/data/test-suites/hybrid-fn-cant-resolve-test-run/testfile.js');
+        it('Should raise error if ClientFunction function not able to resolve test run', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        it('Should raise an error if ClientFunction's function is unable to resolve the test run', function () {
```

File: test/server/compiler-test.js [hunk 30]
```diff
@@ -610,8 +611,8 @@ describe('Compiler', function () {
                 });
         });
 
-        it('Should raise error if Hybrid function uses async function', function () {
-            var testfile = resolve('test/server/data/test-suites/async-function-in-hybrid/testfile.js');
+        it('Should raise error if ClientFunction function uses async function', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        it('Should raise an error if ClientFunction's function uses an async function', function () {
```

File: test/server/compiler-test.js [hunk 31]
```diff
@@ -637,8 +638,8 @@ describe('Compiler', function () {
                 });
         });
 
-        it('Should raise error if Hybrid function uses generator', function () {
-            var testfile = resolve('test/server/data/test-suites/generator-in-hybrid/testfile.js');
+        it('Should raise error if ClientFunction function uses generator', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        it('Should raise an error if ClientFunction's function uses a generator', function () {
```

File: test/server/compiler-test.js [hunk 32]
```diff
@@ -665,8 +666,8 @@ describe('Compiler', function () {
                 });
         });
 
-        it('Should raise error if Hybrid function dependencies is not an object', function () {
-            var testfile = resolve('test/server/data/test-suites/hybrid-deps-not-object/testfile.js');
+        it('Should raise error if ClientFunction function dependencies is not an object', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        it('Should raise an error if ClientFunction's dependency is not an object', function () {
```

File: test/server/compiler-test.js [hunk 33]
```diff
@@ -679,17 +680,17 @@ describe('Compiler', function () {
...
                                   '   4 |\n' +
-                                  " > 5 |var selectYo = Hybrid(() => document.querySelector('#yo'), '42');\n"
+                                  " > 5 |var selectYo = ClientFunction(() => document.querySelector('#yo'), '42');\n"
                     });
                 });
         });
 
-        it('Should raise error if Hybrid function dependency is not a Hybrid function', function () {
-            var testfile = resolve('test/server/data/test-suites/hybrid-dep-not-hybrid/testfile.js');
+        it('Should raise error if ClientFunction function dependency is not a hybrid function', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        it('Should raise an error if ClientFunction's dependency is not a hybrid function', function () {
```

**Comment by VasilyStrelyaev** [comment]
\r-

File: src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js (hunk 4 above)
Comment (inikulin) [inline src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js]: It's should be `IE doesn't have `children` for non-element nodes`

File: src/errors/test-run/templates.js (hunk 11 above)
Comment (inikulin) [inline src/errors/test-run/templates.js]: Let's keep it as is for now. I'll do work on errors later

**Comment by inikulin** [comment]
FPR. Note: I've decided to add base class for ClientFunction and Selector

**Comment by inikulin** [comment]
OK, I'll address test names issues later. Please, review functionality in the meantime

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by inikulin** [comment]
Reverted back to version without base class

**Comment by helen-dikareva** [comment]
lgtm

**Comment by inikulin** [comment]
FPR

**Comment by inikulin** [comment]
Just unstable client test

**Comment by inikulin** [comment]
ping @VasilyStrelyaev @helen-dikareva @AlexanderMoskovkin

File: test/server/compiler-test.js [hunk 34]
```diff
@@ -580,8 +581,8 @@ describe('Compiler', function () {
         });
 
 
-        it('Should raise error if Hybrid function not able to resolve test run', function () {
-            var testfile = resolve('test/server/data/test-suites/hybrid-fn-cant-resolve-test-run/testfile.js');
+        it('Should raise an error if ClientFunction function not able to resolve test run', function () {
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]: `ClientFunction function` x 5

**Comment by VasilyStrelyaev** [comment]
and lgtm

**Comment by inikulin** [comment]
FPR

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by inikulin** [comment]
Yet another unstable client test. Can be merged

**Comment by helen-dikareva** [comment]
lgtm
